### ISSUE TITLE
Truncating large InSpec reports

### ISSUE THREAD

**Issue by alexpop** [chef-cookbooks/audit#422]
There are cases where InSpec reports exceed **10** or even **20 MB** in size. Chef Automate has a hard GRPC ingestion limit of **4 MB**.

I've got a hold of a few of these large reports to understand where things get out of control. I identified two cases that lead to such large reports:

### 1. The `message` field of a control result contains a large text.

Below is a control I extracted from a **9.4 MB** report. The `message` field of this control was **8 MB** in size:
```json
{
  "id": "xccdf_org.cisecurity.benchmarks_rule_1.8_Ensure_updates_patches_and_additional_security_software_are_installed",
  "title": "Ensure updates, patches, and additional security software are installed",
  "results": [
    {
      "status": "failed",
      "code_desc": "Command: `yum check-update --security` stdout is expected to include \"No packages needed for security\"",
      "run_time": 74.522277202,
      "start_time": "2020-02-12T08:27:34+00:00",
      "message": "!!!!!!!!! TEXT_OF_8MB_IN_SIZE_HERE !!!!!!!!!"
    }
  ]
}
```

Attaching here ([message_23k.txt](https://github.com/chef-cookbooks/audit/files/4523473/message_23k.txt)) the first **23 KB** of the **8 MB** `message`. Without this large diff message, this report would be ingested. Even if we didn't have the **8 MB** ingestion limit, storing this large text, processing during API calls and rendering it in the UI would be very challenging.

My proposal is to truncate such messages to a reasonable size that would still allow someone to understand the control failure.
This size can have a default value in the audit cookbook attributes, for example maximum 5000 characters for the `message` field. Is this a reasonable default? 
I'm attaching the `message` field from above truncated at 5000 characters here ([message_5k.txt](https://github.com/chef-cookbooks/audit/files/4523458/message_5k.txt))

---

### 2. A control contains a large number of results

This can happen when a control has loops to check arrays of resources (users, files, servers, containers, images, buckets, etc). For example:

```json
{
  "id": "cis-gcp-benchmark-vms-4.6",
  "title": "Ensure that IP forwarding is not enabled on Instances",
  "results": [
    {
        "status": "failed",
        "code_desc": "Instance srv-pi-gts-a6d61c16-ds0z can_ip_forward is expected to equal false",
        "run_time": 0.000226681,
        "start_time": "2020-04-08T20:49:51+00:00",
        "message": "\nexpected false\n     got true\n"
    },
    {
        "status": "failed",
        "code_desc": "Instance srv-pi-gts-a6d61c16-f069 can_ip_forward is expected to equal false",
        "run_time": 0.000119824,
        "start_time": "2020-04-08T20:49:51+00:00",
        "message": "\nexpected false\n     got true\n"
    },
    {
        "status": "failed",
        "code_desc": "Instance srv-pi-gts-a6d61c16-hb0f can_ip_forward is expected to equal false",
        "run_time": 0.000146165,
        "start_time": "2020-04-08T20:49:51+00:00",
        "message": "\nexpected false\n     got true\n"
    },
    ...
  ]
}
```

The control above is from an **11 MB** report generated by scanning a GCE account. This report has four controls that exceed **5000** results each.
Here's a list of the controls sorted descending by the results count:
```
5788      Ensure that IP forwarding is not enabled on Instances
5776      Ensure "Block Project-wide SSH keys" enabled for VM instances
5762      Ensure that IP forwarding is not enabled on Instances
5762      Ensure that instances are not configured to use the default service account with full access to all Cloud APIs
1783      Ensure "Enable connecting to serial ports" is not enabled for VM Instance
 614      Ensure that Cloud Storage bucket is not anonymously or publicly accessible
 345      Ensure that SSH access is restricted from the internet
 345      Ensure that RDP access is restricted from the internet
  94      Ensure user-managed/external keys for service accounts are rotated every 90 days or less.
  94      User managed service account should not have user managed keys.
  86      Ensure that logging is enabled for Cloud storage buckets
  66      Use corporate login credentials instead of Gmail accounts.
  20      Ensure that ServiceAccount has no Admin privileges.
   7      Ensure legacy networks does not exists for a project
   3      Ensure that Cloud Audit Logging is configured properly across all services and all users from a project
   2      Ensure that the project exists and is in lifecycle state "ACTIVE".
   2      Ensure that object versioning is enabled on log-buckets
...
```

If I truncate the number of results per control at max **50**, essentially this:
```
$ ruby sort.rb
50      Ensure that IP forwarding is not enabled on Instances
50      Ensure "Block Project-wide SSH keys" enabled for VM instances
50      Ensure that IP forwarding is not enabled on Instances
50      Ensure that instances are not configured to use the default service account with full access to all Cloud APIs
50      Ensure "Enable connecting to serial ports" is not enabled for VM Instance
50      Ensure that Cloud Storage bucket is not anonymously or publicly accessible
50      Ensure that SSH access is restricted from the internet
50      Ensure that RDP access is restricted from the internet
50      Ensure user-managed/external keys for service accounts are rotated every 90 days or less.
50      User managed service account should not have user managed keys.
50      Ensure that logging is enabled for Cloud storage buckets
50      Use corporate login credentials instead of Gmail accounts.
20      Ensure that ServiceAccount has no Admin privileges.
 7      Ensure legacy networks does not exists for a project
 3      Ensure that Cloud Audit Logging is configured properly across all services and all users from a project
 2      Ensure that the project exists and is in lifecycle state "ACTIVE".
 2      Ensure that object versioning is enabled on log-buckets
...
```

the initial **11 MB** report is reduced to **400 KB**.

Truncating results is not as straight forward. InSpec does not provide an overall run status (`failed`/`passed`/`skipped`) of a control.
This status is calculated in Automate by traversing all the results for each control. Taking for example the control above with 5788 results, let's say 30 of them failed, 30 of them were skipped and the remaining 5728 passed. If we were to take a slice of the first 50 results and truncate the rest, it's likely that all of those will be passed. If automate is only getting the 50 passed results for this control, it will conclude that the control `passed`, when in fact it `failed` from InSpec's perspective. In order to avoid this, the slicing will have to take into account the statuses of the results and keep failed and skipped results in. In this example, in the 50 results kept, 30 will be failed and 20 skipped. Optionally we could return a summary of the truncated results, something like: `"truncated": {"failed": 0, "skipped": 10, "passed": 5728}`

If we introduce this max results per control attribute in the audit cookbook, is 50 a reasonable default for it?

---

I talked with the InSpec team about this. Depending on how these truncating options evolve, they might make it into InSpec core later. One benefit can be the reduction in time it takes to generate a report. For example, after a control has 50 failed results, don't bother checking the remaining 5738. But this requires major refactoring of the InSpec rspec code.

Any other suggestions or concerns?

**Comment by sean-horn** [comment]
Thanks for going over this in detail, @alexpop 

I think if we can imagine scenarios where truncation of the output of 2. would change the testing results significantly and result in false positives/negatives, then the most correct way that retains the intent of the control result is the way to go.

The two options you mention seem to fit that bill and retain the integrity of the testing.

> In order to avoid this, the slicing will have to take into account the statuses of the results and keep failed and skipped results in. In this example, in the 50 results kept, 30 will be failed and 20 skipped. Optionally we could return a summary of the truncated results, something like: "truncated": {"failed": 0, "skipped": 10, "passed": 5728}

As for keeping 50 results. The number seems good. 5 is probably too small to get a decent picture, 100 is probably too many if we're trying to cut down on overall size of the reports.

**Comment by vjeffrey** [comment]
decision from @kekaichinose :
1 - will be done in inspec
2 - will be done in audit cookbook
3 - metasearch profile stripping will be added to the automate reporter

**Comment by vjeffrey** [comment]
- need to add the metasearch logic to chef-server as well
- need to add results truncation to scan jobs reporting logic
- need to add results truncation to effortless workflow
- some small automate api and ui work to let customer know about truncation

**Comment by alexpop** [comment]
* `ap/truncate-message` is a draft branch where **case 1** is implemented. I'll open a PR for it when [this InSpec merge](https://github.com/inspec/inspec/pull/4994) makes it into an inspec release that is bundled with chef infra client.

 * `ap/truncate-results` is a draft branch where **case 2** is implemented. I'll open a PR from this branch next week.

**Comment by alexpop** [comment]
PR https://github.com/chef-cookbooks/audit/pull/424 is now merged for **case 1** with a default message truncation limit of 10000 characters. Inspec resource stacktraces are also removed based on a new inspec runner option. This will further reduce the report sizes. This has been released today in audit cookbook version `9.3.0`

PR https://github.com/chef-cookbooks/audit/pull/426 is addressing **case 2** described in this issue by limiting the number of results per control at 50 by default. Once this gets the approvals, I'll merge and release another audit cookbook version.

**Comment by alexpop** [comment]
Pull requests #424 and #426 are both merged now and available in the 9.4.0 release of the audit cookbook. Thank you for the feedback!
